lib/tpm/tpm_simulator_test.go

```diff
@@ -0,0 +1,230 @@
+	return nil, nil
+}
+
+func writeEKCertToTPM(t *testing.T, sim *tpmsimulator.Simulator, data []byte) {
+	const nvramRSAEKCertIndex = 0x1c00002
+	err := tpm2.NVDefineSpace(
+		sim,
+		tpm2.HandlePlatform, // Using Platform Authorization.
+		nvramRSAEKCertIndex,
+		"", // As this is the simulator, there isn't a password for Platform Authorization.
+		"", // We do not configure a password for this index. This allows it to be read using the NV index as the auth handle.
+		nil,
+		tpm2.AttrPPWrite| // Allows this NV index to be written with platform authorization.
+			tpm2.AttrPPRead| // Allows this NV index to be read with platform authorization.
+			tpm2.AttrPlatformCreate| // Marks this index as created by the Platform
+			tpm2.AttrAuthRead, // Allows the nv index to be used as an auth handle to read itself.
+
+		uint16(len(data)),
+	)
+	require.NoError(t, err)
+
+	err = tpm2.NVWrite(
+		sim,
+		tpm2.HandlePlatform,
+		nvramRSAEKCertIndex,
+		"",
+		data,
+		0,
+	)
+	require.NoError(t, err)
+}
+
+// To include tests based on this simulator, use the `tpmsimulator` build tag.
+// This requires openssl libraries to be installed on the machine and findable
+// by the compiler. On macOS:
+// brew install openssl
+// export C_INCLUDE_PATH="$(brew --prefix openssl)/include"
+// export LIBRARY_PATH="$(brew --prefix openssl)/lib"
+// go test ./lib/tpm -run TestWithSimulator -tags tpmsimulator
+
+func TestWithSimulator(t *testing.T) {
+	t.Parallel()
+	ctx := context.Background()
+	log := slog.Default()
+
+	ca := &x509.Certificate{
+		SerialNumber:          big.NewInt(1),
+		BasicConstraintsValid: true,
+		IsCA:                  true,
+		NotBefore:             time.Now().Add(-time.Hour),
+		NotAfter:              time.Now().Add(time.Hour),
+	}
+	caPrivKey, err := rsa.GenerateKey(rand.Reader, 4096)
+	require.NoError(t, err)
+	caBytes, err := x509.CreateCertificate(
+		rand.Reader, ca, ca, &caPrivKey.PublicKey, caPrivKey,
+	)
+	require.NoError(t, err)
+	caPEM := pem.EncodeToMemory(
+		&pem.Block{Type: "CERTIFICATE", Bytes: caBytes},
+	)
+
+	sim, err := tpmsimulator.GetWithFixedSeedInsecure(0)
+	require.NoError(t, err)
+	// This is the EKPubHash that results from the EK generated with the seed 0.
+	wantEKPubHash := "1b5bbe2e96054f7bc34ebe7ba9a4a9eac5611c6879285ceff6094fa556af485c"
+
+	attestTPM, err := attest.OpenTPM(&attest.OpenConfig{
+		TPMVersion:     attest.TPMVersion20,
+		CommandChannel: &fakeCmdChannel{ReadWriteCloser: sim},
+	})
+	require.NoError(t, err)
+	t.Cleanup(func() {
+		assert.NoError(t, attestTPM.Close())
+	})
+
+	t.Run("Without EKCert", func(t *testing.T) {
+		queryRes, attParams, solve, _, err := attestWithTPM(ctx, log, attestTPM)
+		require.NoError(t, err)
+
+		// Check QueryRes looks right.
+		require.Equal(t, wantEKPubHash, queryRes.EKPubHash)
+		require.NotEmpty(t, queryRes.EKPub)
+		require.False(t, queryRes.EKCertPresent)
+		require.Empty(t, queryRes.EKCertSerial)
+		require.Empty(t, queryRes.EKCert)
```

Comment:
```suggestion
		assert.Equal(t, wantEKPubHash, queryRes.EKPubHash)
		assert.NotEmpty(t, queryRes.EKPub)
		assert.False(t, queryRes.EKCertPresent)
		assert.Empty(t, queryRes.EKCertSerial)
		assert.Empty(t, queryRes.EKCert)
```

https://go.dev/wiki/TestComments#keep-going

Reply:
018d4e0419